lib/api/xmlrpc/v1/xmlrpc.class.php

```diff
@@ -2687,9 +2690,73 @@ public function getTestCasesForTestPlan($args) {
             }
         }
 
+        // Do we need the custom fields?
+        if (! is_null( $recordset ) && ($opt[self::$customFieldsParamName] || is_array($opt[self::$customFieldsParamName]))) {
+            $itemSet = array_keys( $recordset );
+            switch($options['output']) {
+                case 'mapOfArray' :
+                case 'mapOfMap' :
+                    foreach( $itemSet as $itemKey ) {
+                        $keySet = array_keys( $recordset[$itemKey] );
+                        $target = &$recordset[$itemKey];
+                        foreach( $keySet as $accessKey ) {
+                            $target[$accessKey]['customfields'] = $this->_testcaseCustomFieldData(
+                                $target[$accessKey], $tplanInfo['parent_id'], $opt[self::$customFieldsParamName]);
+                        }
+                    }
+                    break;
+
+                case 'array' :
+                case 'map' :
+                default :
+                    foreach( $itemSet as $accessKey ) {
+                        $recordset[$accessKey]['customfields'] = $this->_testcaseCustomFieldData(
+                            $recordset[$accessKey], $tplanInfo['parent_id'], $opt[self::$customFieldsParamName]);
+                    }
+                    break;
+            }
+        }
+
         return $recordset;
     }
 
+    /**
+     * Get a map of all custom fields from the given testcase.
+     *
+     * This can be filtered by providing an array with the allowed
+     * custom field names as $cf_options.
+     *
+     * @param mixed $testcase
+     *              The testcase to query for.
+     * @param int $project_id
+     *              The project to query for.
+     * @param array $cf_options
+     *              The optional name filter.
+     * @return array
+     * @access private
+     */
+    private function _testcaseCustomFieldData(&$testcase, $project_id, &$cf_options) {
+        $cf = $this->tcaseMgr->get_linked_cfields_at_design(
+            $testcase['tcase_id'],$testcase['tcversion_id'],
+            null,null, $project_id);
+        $filtered_cf = array();
+
+        foreach ($cf as $cf_id => $cfield) {

```

Comment:
please extract the invariant from the loop  => is_array($cf_options), do the check using the function call to is_array() just once outside the loop saving result in a variable